### ISSUE TITLE
[BUG] Azure Search SDK does not seem to respect cancellation

### ISSUE THREAD

**Issue by mattgotteiner** [Azure/azure-sdk-for-net#33369]
### Library name and version

Azure.Search.Documents 11.4.0

### Describe the bug

We are passing along the CancellationToken properly to the search SDK, but it didn't seem to respect the cancellation times. Our services are set to timeout the HTTP request from our browser to our backend service at 15 seconds. This terminates the HTTP connection and should cascade cancellations throughout the stack, including azure search. However, we found that it didn't happen. In fact, some search and indexing requests took 1-5 minutes to timeout.

In our logs for these requests our cancellation tokens were canceled at the 15-second mark. However, we see the elapsed time was well beyond 15 seconds.

### Expected behavior

When cancelling a search or indexing request, this cancellation goes to the search service

### Actual behavior

```
System.AggregateException: Retry failed after 4 tries. Retry settings
 can be adjusted in ClientOptions.Retry or by configuring a custom retry policy in ClientOptions.RetryPolicy. (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.)
 (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was canceled.) ---> System.Threading.Tasks.TaskCanceledException: The operation was cancelled
 because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout. ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled. ---> System.IO.IOException: Unable to read data from
 the transport connection: Operation canceled. ---> System.Net.Sockets.SocketException (125): Operation canceled --- End of inner exception stack trace --- at void System.Net.Sockets.Socket+AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken
 cancellationToken) at async ValueTask<int> System.Net.Security.SslStream.EnsureFullTlsFrameAsync<TIOAdapter>(TIOAdapter adapter) at async ValueTask<int> System.Net.Security.SslStream.ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, Memory<byte> buffer) at
 async ValueTask System.Net.Http.HttpConnection.InitialFillAsync(bool async) at async Task<HttpResponseMessage> System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, bool async, CancellationToken cancellationToken) --- End of inner exception
 stack trace --- at void System.Net.Http.HttpClient.HandleFailure(Exception e, bool telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts) at async Task<HttpResponseMessage>
 System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request)+Core(?) at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async) at async ValueTask Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage
 message, ReadOnlyMemory<HttpPipelinePolicy> pipeline) at async ValueTask Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async) at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage
 message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async) --- End of inner exception stack trace --- at void Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken,
 Exception inner, TimeSpan timeout) at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async) at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage
 message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async) at async ValueTask Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async) at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage
 message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async) --- End of inner exception stack trace --- at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async) at async
 Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchInternal<T>(SearchOptions options, string operationName, bool async, CancellationToken cancellationToken) x 2 at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchAsync<T>(string
 searchText, SearchOptions options, CancellationToken cancellationToken)
```

Received a 503 response after 207 seconds (should have been canceled at 15 seconds...)

### Reproduction Steps

Call cancel on a search sdk method

### Environment

Windows .NET 6

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by ShivangiReja** [comment]
Hi @mattmsft, I tried to add `CancellationToken` in Search operation and canceled it after 1 second and it worked as expected. Attaching the test and logs for your reference:
![image](https://user-images.githubusercontent.com/45216704/212436870-ee53e59f-9eb7-46d1-8c99-90a91b6d3e69.png)

**Comment by ccurrens** [comment]
@ShivangiReja, let me share a few more details about this bug report. This issue was found while there was an availability issue with our search service. All of our replicas were experiencing issues related to high memory (and possibly high CPU as well), and so there were general connection issues. Under normal circumstances, cancellation works great. But under these failure circumstances, cancellation was not being respected. This negatively affects our entire application, even the parts that don't use azure search, because connections are left open by the SDK.

The way cancellation in our app works is that it is tied to ASP.NET request timeout, which is 15 seconds. This is controlled by a cancellation token that we pass along to all external services, including Azure Search. We actually saw three different stacks, but I suspect they are likely all related. I can confirm in all of these stacks that our cancellation token was canceled at 15 seconds, however these exceptions occurred hundreds of seconds later, well after it should have been canceled.

1. This is the one that @mattmsft posted, which seems to happen during send on an existing connection (_possibly_ a connection that was just opened, but I can't confirm that). Took 300 seconds to fail, although the passed-in cancelation token was canceled at 15 seconds)

```
System.AggregateException: Retry failed after 4 tries. Retry settings can be adjusted in ClientOptions.Retry or by configuring a custom retry policy in ClientOptions.RetryPolicy.
(The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.)
(The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.)
(The operation was canceled.)
    ---> System.Threading.Tasks.TaskCanceledException: The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.
    ---> System.Threading.Tasks.TaskCanceledException: The operation was canceled.
    ---> System.IO.IOException: Unable to read data from the transport connection: Operation canceled.
    ---> System.Net.Sockets.SocketException (125): Operation canceled
    --- End of inner exception stack trace ---
        at void System.Net.Sockets.Socket+AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
        at async ValueTask<int> System.Net.Security.SslStream.EnsureFullTlsFrameAsync<TIOAdapter>(TIOAdapter adapter)
        at async ValueTask<int> System.Net.Security.SslStream.ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, Memory<byte> buffer)
        at async ValueTask System.Net.Http.HttpConnection.InitialFillAsync(bool async)
        at async Task<HttpResponseMessage> System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
    --- End of inner exception stack trace ---
        at void System.Net.Http.HttpClient.HandleFailure(Exception e, bool telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
        at async Task<HttpResponseMessage> System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request)+Core(?)
        at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async)
        at async ValueTask Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
        at async ValueTask Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
    --- End of inner exception stack trace ---
        at void Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
        at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async ValueTask Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
    --- End of inner exception stack trace ---
        at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchInternal<T>(SearchOptions options, string operationName, bool async, CancellationToken cancellationToken) x 2
        at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchAsync<T>(string searchText, SearchOptions options, CancellationToken cancellationToken)
```

2.  Received a 503 response after 207 seconds, even though it was canceled at 15 seconds.

```
Azure.RequestFailedException: The service is too busy
    Status: 503 (Service Unavailable)
    Content: {"error":{"code":"","message":"The service is too busy"}}
    Headers:
        Cache-Control: no-store, no-cache
        Pragma: no-cache
        Server: Microsoft-IIS/10.0
        client-request-id: a7a39c8a-19a6-44d9-b1d0-bdd2c9d75b7f
        x-ms-client-request-id: a7a39c8a-19a6-44d9-b1d0-bdd2c9d75b7f
        request-id: a7a39c8a-19a6-44d9-b1d0-bdd2c9d75b7f
        elapsed-time: 3528
        Strict-Transport-Security: REDACTED
        Date: Mon, 05 Dec 2022 15:06:58 GMT
        Content-Length: 57
        Content-Type: application/json; charset=utf-8
        Content-Language: REDACTED
        Expires: -1
at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchInternal<T>(SearchOptions options, string operationName, bool async, CancellationToken cancellationToken) x 2
at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchAsync<T>(string searchText, SearchOptions options, CancellationToken cancellationToken)
```

3. During initial connection to the service (300 seconds, although canceled at 15 seconds)

```
System.AggregateException: Retry failed after 4 tries. Retry settings can be adjusted in ClientOptions.Retry or by configuring a custom retry policy in ClientOptions.RetryPolicy. (Operation timed out ([REDACTED].search.windows.net:443)) (Operation timed out ([REDACTED].search.windows.net:443)) (The operation was canceled.)
    ---> Azure.RequestFailedException: Operation timed out ([REDACTED].search.windows.net:443)
    ---> System.Net.Http.HttpRequestException: Operation timed out ([REDACTED].search.windows.net:443)
    ---> System.Net.Sockets.SocketException (110): Operation timed out at void System.Net.Sockets.Socket+AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
        at async Task System.Net.Sockets.Socket.ConnectAsync(EndPoint remoteEP)+WaitForConnectWithCancellation(?)
        at async ValueTask<ValueTuple<Socket, Stream>> System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
    --- End of inner exception stack trace ---
        at async ValueTask<ValueTuple<Socket, Stream>> System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
        at async ValueTask<ValueTuple<Socket, Stream, TransportContext>> System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
        at async ValueTask<HttpConnection> System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
        at async Task System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage request)
        at async ValueTask<T> System.Threading.Tasks.TaskCompletionSourceWithCancellation<T>.WaitWithCancellationAsync(CancellationToken cancellationToken)
        at async ValueTask<HttpConnection> System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
        at async ValueTask<HttpResponseMessage> System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, bool async, bool doRequestAuth, CancellationToken cancellationToken)
        at async ValueTask<HttpResponseMessage> System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
        at async Task<HttpResponseMessage> System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request)+Core(?)
        at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async)
    --- End of inner exception stack trace ---
        at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async)
        at async ValueTask Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
        at async ValueTask Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async ValueTask Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
    --- End of inner exception stack trace ---
        at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
        at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchInternal<T>(SearchOptions options, string operationName, bool async, CancellationToken cancellationToken) x 2
        at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchAsync<T>(string searchText, SearchOptions options, CancellationToken cancellationToken)
```

For 1 and 3, I wonder if the runtime is passing along the `CancellationToken` to the `HttpConnection`. Thankfully it eventually fails when it hits the `NetworkTimeout` on the connection/socket. Perhaps an easy solution for these would be to allow callers to specify a timeout, which can be passed to the connection or socket. That way, _around_ the same time the token is canceled, any outstanding requests would timeout as well. It's not ideal, but it's better than what is currently happening.

For 2, I wish there was more stack trace, but that's all I got from the SDK (maybe that's a separate issue?). Since it actually returned data (a 503 from the service), it must have been that the task was waiting on a response from the service. I would imagine that the cancellation token would be passed along in the `HttpClient`, so I don't understand why it wasn't canceled.

So, like was mentioned, this really only happens when there are connection and communication issues between the client and server. I have found in the past that using a tool like [Clumsy](http://jagt.github.io/clumsy/) to introduce lag and dropped packets is very helpful to reproduce issues like this.

**Comment by KrzysztofCwalina** [comment]
I wonder if it's not related to https://github.com/dotnet/runtime/issues/70656

**Comment by JoshLove-msft** [comment]
Awesome find.. It looks like this was fixed in https://github.com/dotnet/runtime/pull/65239 which would be included in .NET 7.

**Comment by ccurrens** [comment]
I would want confirmation on that, because I don't see the cancellation token being passed along into `HttpConnection.InitialFillAsync`, which calls into the `SslStream`, so I'm not sure what cancellation token would be used in `EnsureFullTlsFrameAsync`.

I am also wondering if the third stack trace would be fixed with this as well, because I notice that it doesn't go down the path of `EnsureFullTlsFrameAsync`. That one may be because I seems that HttpConnection uses [a separate cancellation token source during connection](https://github.com/dotnet/runtime/blob/c43506135b6191b53b980b7e754ad08df46721e8/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L483) rather than the one passed along through the `HttpClient`.

**Comment by stephentoub** [comment]
> this is a .NET 6 timeframe version:
> https://github.com/dotnet/runtime/blob/2ac023c8e9df61fe4557212cbd13956042fb47c5/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs
> It does not take the cancellation token.

It does, it's just more obscured that it does.  That method:
https://github.com/dotnet/runtime/blob/2ac023c8e9df61fe4557212cbd13956042fb47c5/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs#L826
takes a struct argument, a TIOAdapter, and its ReadAsync method is called:
https://github.com/dotnet/runtime/blob/2ac023c8e9df61fe4557212cbd13956042fb47c5/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs#L844
That adapter's ReadAsync method is:
https://github.com/dotnet/runtime/blob/2ac023c8e9df61fe4557212cbd13956042fb47c5/src/libraries/System.Net.Security/src/System/Net/Security/ReadWriteAdapter.cs#L33-L34
which you can see does pass in the token that was given to its ctor. And you can see the adapter being constructed with a token in all the places one is created, e.g.
https://github.com/dotnet/runtime/blob/2ac023c8e9df61fe4557212cbd13956042fb47c5/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs#L203

**Comment by stephentoub** [comment]
> I don't see the cancellation token being passed along into HttpConnection.InitialFillAsync

It's a bit convoluted, but for historical reasons, we don't currently pass the cancellation token through into calls as part of setting up a connection.  Instead, we register with the token once:
https://github.com/dotnet/runtime/blob/88dd8c77b155e3a30db68bc212b11b2fc67298b7/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L508
to tear down the stream if/when cancellation is requested.  There's an issue somewhere in dotnet/runtime which my searching skills can't find at the moment that tracks switching the implementation over to instead thread the token through.  That's more of a code cleanup thing, though; I'm not aware of known bugs resulting from the current approach.

**Comment by stephentoub** [comment]
cc: @mihazupan

**Comment by MihaZupan** [comment]
I'm not familiar with the codebase in this repo, but digging through the stack a bit:
- The exception message that you observed was
  - > The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.
- As far as I can tell, this message will only be produced by [`ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout`](https://github.com/Azure/azure-sdk-for-net/blob/e04f2ee2108f43c08d04f527440e2e53626f95cc/sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs#L166-L179)
- Before that exception can be thrown, you call [`CancellationHelper.ThrowIfCancellationRequested(originalToken);`](https://github.com/Azure/azure-sdk-for-net/blob/e04f2ee2108f43c08d04f527440e2e53626f95cc/sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs#L169)
- The `originalToken` is the `oldToken` in `ProcessAsync` above, which comes from `message.CancellationToken`

**So for you to see this exception message, the `HttpMessage.CancellationToken` couldn't have been signaled.**

That is, the cancellation token that was given to HttpClient wasn't signaled (except by the eventual timeout). From the looks of it, you have a cancellation issue higher up in the stack, before you ever call into HttpClient.

**Comment by JoshLove-msft** [comment]
> So for you to see this exception message, the HttpMessage.CancellationToken couldn't have been signaled.

Good point - if the originalToken was signaled, we wouldn't see the exception message related to the timeout. 

@ccurrens can you please include a code snippet demonstrating the scenario?

**Comment by ccurrens** [comment]
@MihaZupan Thanks for the in-depth look at the SDK code. From looking at our code, we simply pass the `CancellationToken` from the ASP.NET action (I believe it's tied to or is simply `HttpContext.RequestAborted`). We have logs that the client request was aborted in 15 seconds, which should have triggered `RequestAborted`, which should have cancelled the search query.

I went back and I found logs that show that during the service interruption, cancellation was working properly for a subset of requests. This one was quickly abandoned after 12ms (probably a user navigating away):

```
System.Threading.Tasks.TaskCanceledException: The operation was canceled.
   at void Azure.Core.CancellationHelper.ThrowOperationCanceledException(Exception innerException, CancellationToken cancellationToken)
   at void Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
   at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async) x 2
   at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchInternal<T>(SearchOptions options, string operationName, bool async, CancellationToken cancellationToken) x 2
   at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchAsync<T>(string searchText, SearchOptions options, CancellationToken cancellationToken)
```

This one was canceled by the client at our 15 second configured timeout, which propagated to the backend service, and canceled the search request immediately at 15 seconds. From the stack, it used `oldToken`, since the cancellation exception comes from `Azure.Core.CancellationHelper.ThrowOperationCanceledException`.

```
System.AggregateException: Retry failed after 2 tries. Retry settings can be adjusted in ClientOptions.Retry or by configuring a custom retry policy in ClientOptions.RetryPolicy. (Operation timed out ([REDACTED].search.windows.net:443)) (The operation was canceled.)
 ---> Azure.RequestFailedException: Operation timed out ([REDACTED].search.windows.net:443)
 ---> System.Net.Http.HttpRequestException: Operation timed out ([REDACTED].search.windows.net:443)
 ---> System.Net.Sockets.SocketException (110): Operation timed out
   at void System.Net.Sockets.Socket+AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at async Task System.Net.Sockets.Socket.ConnectAsync(EndPoint remoteEP)+WaitForConnectWithCancellation(?)
   at async ValueTask<ValueTuple<Socket, Stream>> System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at async ValueTask<ValueTuple<Socket, Stream>> System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
   at async ValueTask<ValueTuple<Socket, Stream, TransportContext>> System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
   at async ValueTask<HttpConnection> System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
   at async Task System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage request)
   at async ValueTask<T> System.Threading.Tasks.TaskCompletionSourceWithCancellation<T>.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at async ValueTask<HttpConnection> System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
   at async ValueTask<HttpResponseMessage> System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, bool async, bool doRequestAuth, CancellationToken cancellationToken)
   at async ValueTask<HttpResponseMessage> System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
   at async Task<HttpResponseMessage> System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request)+Core(?)
   at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async)
   --- End of inner exception stack trace ---
   at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async)
   at async ValueTask Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
   at async ValueTask Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   --- End of inner exception stack trace ---
   at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchInternal<T>(SearchOptions options, string operationName, bool async, CancellationToken cancellationToken) x 2
   at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchAsync<T>(string searchText, SearchOptions options, CancellationToken cancellationToken)
 ---> (Inner Exception #1) System.Threading.Tasks.TaskCanceledException: The operation was canceled.
   at void Azure.Core.CancellationHelper.ThrowOperationCanceledException(Exception innerException, CancellationToken cancellationToken)
   at void Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
   at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
```

Here's another one where I can see that the task was canceled by the client at 15 seconds, and the `RequestAborted` token was canceled. The stack trace also confirms that `originalToken` was the source of the cancellation, which from our logs shows as 15 seconds, but it still took 296 seconds for some reason.

```
System.AggregateException: Retry failed after 3 tries. Retry settings can be adjusted in ClientOptions.Retry or by configuring a custom retry policy in ClientOptions.RetryPolicy. (Operation timed out ([REDACTED].search.windows.net:443)) (The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.) (The operation was canceled.)
 ---> Azure.RequestFailedException: Operation timed out ([REDACTED].search.windows.net:443)
 ---> System.Net.Http.HttpRequestException: Operation timed out ([REDACTED].search.windows.net:443)
 ---> System.Net.Sockets.SocketException (110): Operation timed out
   at void System.Net.Sockets.Socket+AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at async Task System.Net.Sockets.Socket.ConnectAsync(EndPoint remoteEP)+WaitForConnectWithCancellation(?)
   at async ValueTask<ValueTuple<Socket, Stream>> System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at async ValueTask<ValueTuple<Socket, Stream>> System.Net.Http.HttpConnectionPool.ConnectToTcpHostAsync(string host, int port, HttpRequestMessage initialRequest, bool async, CancellationToken cancellationToken)
   at async ValueTask<ValueTuple<Socket, Stream, TransportContext>> System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
   at async ValueTask<HttpConnection> System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
   at async Task System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(HttpRequestMessage request)
   at async ValueTask<T> System.Threading.Tasks.TaskCompletionSourceWithCancellation<T>.WaitWithCancellationAsync(CancellationToken cancellationToken)
   at async ValueTask<HttpConnection> System.Net.Http.HttpConnectionPool.GetHttp11ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
   at async ValueTask<HttpResponseMessage> System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, bool async, bool doRequestAuth, CancellationToken cancellationToken)
   at async ValueTask<HttpResponseMessage> System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
   at async Task<HttpResponseMessage> System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request)+Core(?)
   at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async)
   --- End of inner exception stack trace ---
   at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async)
   at async ValueTask Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
   at async ValueTask Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   --- End of inner exception stack trace ---
   at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchInternal<T>(SearchOptions options, string operationName, bool async, CancellationToken cancellationToken) x 2
   at async Task<Response<SearchResults<T>>> Azure.Search.Documents.SearchClient.SearchAsync<T>(string searchText, SearchOptions options, CancellationToken cancellationToken)
 ---> (Inner Exception #1) System.Threading.Tasks.TaskCanceledException: The operation was cancelled because it exceeded the configured timeout of 0:01:40. Network timeout can be adjusted in ClientOptions.Retry.NetworkTimeout.
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at void System.Net.Http.HttpClient.HandleFailure(Exception e, bool telemetryStarted, HttpResponseMessage response, CancellationTokenSource cts, CancellationToken cancellationToken, CancellationTokenSource pendingRequestsCts)
   at async Task<HttpResponseMessage> System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request)+Core(?)
   at async ValueTask Azure.Core.Pipeline.HttpClientTransport.ProcessAsync(HttpMessage message, bool async)
   at async ValueTask Azure.Core.Pipeline.HttpPipelineTransportPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
   at async ValueTask Azure.Core.Pipeline.RequestActivityPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   --- End of inner exception stack trace ---
   at void Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
   at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RedirectPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)<---

 ---> (Inner Exception #2) System.Threading.Tasks.TaskCanceledException: The operation was canceled.
   at void Azure.Core.CancellationHelper.ThrowOperationCanceledException(Exception innerException, CancellationToken cancellationToken)
   at void Azure.Core.Pipeline.ResponseBodyPolicy.ThrowIfCancellationRequestedOrTimeout(CancellationToken originalToken, CancellationToken timeoutToken, Exception inner, TimeSpan timeout)
   at async ValueTask Azure.Core.Pipeline.ResponseBodyPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
   at async ValueTask Azure.Core.Pipeline.LoggingPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
```

So, it seems to work sometimes, and other times it doesn't. I had considered this being a threadpool starvation problem, but our metrics don't seem to agree with that, at least I don't see a large amount of queued threads and we had good throughput and low latency for requests that didn't use azure search during the outage. Could it be something to do with `RetryPolicy` not respecting the `CancellationToken` on retries?

**Comment by MihaZupan** [comment]
> Here's another one where I can see that the task was canceled by the client at 15 seconds, and the RequestAborted token was canceled. The stack trace also confirms that originalToken was the source of the cancellation, which from our logs shows as 15 seconds, but it still took 296 seconds for some reason.

From the exception message, there were 3 retries:
- First one timed out after 100 seconds
- Second one timed out after 100 seconds
- Third one was cancelled by the original (presumably your) token

That is, the token you supplied to the operation was not canceled for at least 200+ seconds?

> I can see that the task was canceled by the client at 15 seconds, and the RequestAborted token was canceled

> which from our logs shows as 15 seconds

How are you determining this? Do you have a registration on `HttpContext.RequestAborted` that does the logging?

If this information is coming from the client, could it be that the client went away without properly closing the connection? In that case the server may not realize that for a while, and so it won't signal the RequestAborted.

**Comment by ccurrens** [comment]
> If this information is coming from the client, could it be that the client went away without properly closing the connection? In that case the server may not realize that for a while, and so it won't signal the RequestAborted.

No, we do not register a callback on the aborted request that does logging. I am basing this on other logs I see and making an assumption, so maybe my mental model of how `RequestAborted` works is wrong.

The client in this case is another ASP.NET web app. The request goes from [browser] -> [Service1] -> [Service2]. Communication between [Service1] and [Service2] is done via `HttpClient`. I see in our logs on [Service1] that after 15 seconds the request via HttpClient to [Service2] times out at 15 seconds (it throws an exception which is logged) and we return a 500 failure back to the [browser]. When this happens, we can look into logs on [Service2] and, in the normal case, see that async operations are cancelled immediately -- `TaskCanceledException` is logged around the same millisecond as the cancellation log in [Service2] -- and the status code in our final request log is set to 499 to indicate the request was aborted (we check `RequestAborted.IsCancellationRequested` in an ASP.NET `IExceptionHandler`).

My understanding is that when `HttpClient` times out on [Service1], effectively aborting the request to [Service2], that it would cause `RequestAborted` to be canceled on [Service2]. Mostly because that's what we've always observed as happening. If that's not a guarantee, that's a bit of a shock to me, because AFAIK, `RequestAborted` is the cancellation token that is passed to ASP.NET controllers. To me, it would somewhat reduce the usefulness of the `CancellationToken` in the controller.

Is is possible that `HttpClient` timeout on the client would not cause `RequestAborted` on the server to be canceled in some cases?

**Comment by MihaZupan** [comment]
> My understanding is that when HttpClient times out on [Service1], effectively aborting the request to [Service2], that it would cause RequestAborted to be canceled on [Service2].
> Is is possible that HttpClient timeout on the client would not cause RequestAborted on the server to be canceled in some cases?

Normally that's the case, yes. If a timeout occurs in HttpClient (on HTTP/1.X), we'll tear down the connection and close the socket.
It's up to the OS to communicate to the TCP peer that a connection is being closed.

If, for example, you happened to lose network connectivity at some point, the FIN frames may not reach the server, and the server has its logic to decide when to give up on a non-responsive connection.

If you're using HTTP/2, you can try configuring Keep-Alive pings to shorten that time: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/options?view=aspnetcore-7.0#http2-keep-alive-ping-configuration

Before we spend more time investigating the issue, it'd be great if you could confirm whether the `RequestAborted` was actually signaled in these cases.

**Comment by ccurrens** [comment]
> Before we spend more time investigating the issue, it'd be great if you could confirm whether the RequestAborted was actually signaled in these cases.

I can only tell you that they were signaled, but I can't confirm with data they were done at the 15 second mark. Best I have is observations. We are using HTTP/1.x on alpine linux and it seems unlikely that over the 1h30m period all 50,000 of the requests that took considerably longer than 15 seconds once [Service1] aborted them lost FIN frames.

Best I can do at this point, then, is to add a log for when it's canceled and wait for another outage.

**Comment by ShivangiReja** [comment]
Based on the ongoing discussion, I have moved this issue to the dotnet/runtime repo - https://github.com/dotnet/runtime/issues/98452.
